blockchain/types/tx_internal_data_ethereum_set_code.go

```diff
@@ -0,0 +1,573 @@
+}
+
+func (t *TxInternalDataEthereumSetCode) GetRecipient() *common.Address {
+	return t.Recipient
+}
+
+func (t *TxInternalDataEthereumSetCode) GetAmount() *big.Int {
+	return new(big.Int).Set(t.Amount)
+}
+
+func (t *TxInternalDataEthereumSetCode) GetPayload() []byte {
+	return t.Payload
+}
+
+func (t *TxInternalDataEthereumSetCode) GetAccessList() AccessList {
+	return t.AccessList
+}
+
+func (tx *TxInternalDataEthereumSetCode) GetAuthorizationList() AuthorizationList {
+	return tx.AuthorizationList
+}
+
+func (t *TxInternalDataEthereumSetCode) GetGasTipCap() *big.Int {
+	return t.GasTipCap
+}
+
+func (t *TxInternalDataEthereumSetCode) GetGasFeeCap() *big.Int {
+	return t.GasFeeCap
+}
+
+func (t *TxInternalDataEthereumSetCode) GetHash() *common.Hash {
+	return t.Hash
+}
+
+func (t *TxInternalDataEthereumSetCode) SetHash(h *common.Hash) {
+	t.Hash = h
+}
+
+func (t *TxInternalDataEthereumSetCode) SetSignature(signatures TxSignatures) {
+	if len(signatures) != 1 {
+		logger.Crit("TxTypeEthereum can receive only single signature!")
+	}
+
+	t.V = signatures[0].V
+	t.R = signatures[0].R
+	t.S = signatures[0].S
+}
+
+func (t *TxInternalDataEthereumSetCode) RawSignatureValues() TxSignatures {
+	return TxSignatures{&TxSignature{t.V, t.R, t.S}}
+}
+
+func (t *TxInternalDataEthereumSetCode) ValidateSignature() bool {
+	v := byte(t.V.Uint64())
+	return crypto.ValidateSignatureValues(v, t.R, t.S, false)
+}
+
+func (t *TxInternalDataEthereumSetCode) RecoverAddress(txhash common.Hash, homestead bool, vfunc func(*big.Int) *big.Int) (common.Address, error) {
+	V := vfunc(t.V)
+	return recoverPlain(txhash, t.R, t.S, V, homestead)
+}
+
+func (t *TxInternalDataEthereumSetCode) RecoverPubkey(txhash common.Hash, homestead bool, vfunc func(*big.Int) *big.Int) ([]*ecdsa.PublicKey, error) {
+	V := vfunc(t.V)
+
+	pk, err := recoverPlainPubkey(txhash, t.R, t.S, V, homestead)
+	if err != nil {
+		return nil, err
+	}
+
+	return []*ecdsa.PublicKey{pk}, nil
+}
+
+func (t *TxInternalDataEthereumSetCode) ChainId() *big.Int {
+	return t.ChainID
+}
+
+func (t *TxInternalDataEthereumSetCode) Equal(a TxInternalData) bool {
+	ta, ok := a.(*TxInternalDataEthereumSetCode)
+	if !ok {
+		return false
+	}
+
+	return t.ChainID.Cmp(ta.ChainID) == 0 &&
+		t.AccountNonce == ta.AccountNonce &&
+		t.GasFeeCap.Cmp(ta.GasFeeCap) == 0 &&
+		t.GasTipCap.Cmp(ta.GasTipCap) == 0 &&
+		t.GasLimit == ta.GasLimit &&
+		equalRecipient(t.Recipient, ta.Recipient) &&
+		t.Amount.Cmp(ta.Amount) == 0 &&
+		reflect.DeepEqual(t.AccessList, ta.AccessList) &&
+		reflect.DeepEqual(t.AuthorizationList, ta.AuthorizationList) &&
+		t.V.Cmp(ta.V) == 0 &&
+		t.R.Cmp(ta.R) == 0 &&
+		t.S.Cmp(ta.S) == 0
+}
+
+func (t *TxInternalDataEthereumSetCode) IntrinsicGas(currentBlockNumber uint64) (uint64, error) {
+	gas, err := IntrinsicGas(t.Payload, t.AccessList, t.Recipient == nil, *fork.Rules(big.NewInt(int64(currentBlockNumber))))
+	return gas + uint64(len(t.AuthorizationList))*params.CallNewAccountGas, err
```

Comment:
@Mdaiki0730 Please move this logic into the IntrinsicGas function like `IntrinsicGas(payload, accesslist, authlist, ...)`. Ideally, `TxInternalDataEthereum*.IntrinsicGas` should be a proxy for the global `func IntrinsicGas()`. cc @tnasu

Reply:
There was a plan to remove the subclass. Ok, I'll move it.
Is the reason for removing because there are often duplicates?

Reply:
Exactly. Duplication is the reason. It's easy to miss something when we have 20+ files implementing subclasses.